File: native-schema-registry/golang/pkg/gsrserde-go/core/wire_format.go
package gsrserde

import (
	"errors"
	"fmt"

	"github.com/google/uuid"
)

// Wire-format constants — parity-anchored to Java
// common/src/main/java/com/amazonaws/services/schemaregistry/utils/AWSSchemaRegistryConstants.java.
const (
	// WireFormatVersionByte is the GSR header version. Anchored to
	// AWSSchemaRegistryConstants.HEADER_VERSION_BYTE = (byte) 3.
	WireFormatVersionByte byte = 0x03

	// CompressionByteNone signals an uncompressed payload.
	// AWSSchemaRegistryConstants.COMPRESSION_DEFAULT_BYTE = (byte) 0.
	CompressionByteNone byte = 0x00

	// CompressionByteZlib signals a zlib-compressed payload.
	// AWSSchemaRegistryConstants.COMPRESSION_BYTE = (byte) 5.
	CompressionByteZlib byte = 0x05

	// HeaderVersionByteSize is the size of the version byte. Anchored to
	// AWSSchemaRegistryConstants.HEADER_VERSION_BYTE_SIZE = 1.
	HeaderVersionByteSize = 1

	// CompressionByteSize is the size of the compression byte. Anchored to
	// AWSSchemaRegistryConstants.COMPRESSION_BYTE_SIZE = 1.
	CompressionByteSize = 1

	// SchemaVersionIDSize is the size of the 16-byte UUID. Anchored to
	// AWSSchemaRegistryConstants.SCHEMA_VERSION_ID_SIZE = 16.
	SchemaVersionIDSize = 16

	// WireFormatHeaderSize is the total prefix length: 1 + 1 + 16 = 18 bytes.
	WireFormatHeaderSize = HeaderVersionByteSize + CompressionByteSize + SchemaVersionIDSize
)

// ErrIncompatibleData is the sentinel for payloads whose 18-byte prefix is
// missing or malformed. Anchored to Java GlueSchemaRegistryIncompatibleDataException.
// Callers MUST check with errors.Is.
var ErrIncompatibleData = errors.New("payload is not in GSR wire format")

// EncodeWireFormat produces the 18-byte GSR prefix followed by payload bytes.
//
// Wire layout (Java parity: SerializationDataEncoder.write at
// serializer-deserializer/src/main/java/com/amazonaws/services/schemaregistry/serializers/SerializationDataEncoder.java:54-70):
//
//	byte 0:        0x03                                   (version)
//	byte 1:        0x00 | 0x05                            (compression)
//	bytes 2..9:    schemaVersionID MSB, big-endian        (8 bytes)
//	bytes 10..17:  schemaVersionID LSB, big-endian        (8 bytes)
//	bytes 18..:    payload (already-compressed if compressionByte == 0x05)
//
// payload must already be in its final compressed-or-not form. The caller is
// responsible for compressing before calling this, because the protobuf
// message-index prefix has to be applied before compression (mirrors Java).
//
// Returns ErrIncompatibleData wrapped via fmt.Errorf when compressionByte is
// neither CompressionByteNone nor CompressionByteZlib. UUID parse failures
// surface as wrapped uuid.Parse errors.
func EncodeWireFormat(schemaVersionID string, compressionByte byte, payload []byte) ([]byte, error) {
	if compressionByte != CompressionByteNone && compressionByte != CompressionByteZlib {
		return nil, fmt.Errorf("%w: unsupported compression byte %#x", ErrIncompatibleData, compressionByte)
	}

	u, err := uuid.Parse(schemaVersionID)
	if err != nil {
		return nil, fmt.Errorf("schema version id is not a valid UUID: %w", err)
	}

	out := make([]byte, 0, WireFormatHeaderSize+len(payload))
	out = append(out, WireFormatVersionByte)
	out = append(out, compressionByte)
	out = append(out, u[:]...) // uuid.UUID is already [16]byte in MSB→LSB big-endian layout matching Java's putLong(MSB)+putLong(LSB).
	out = append(out, payload...)

	return out, nil
}

// DecodeWireFormat parses the 18-byte GSR prefix and returns
// (schemaVersionID, compressionByte, payload).
//
// Java parity: GlueSchemaRegistryDeserializerDataParser.getSchemaVersionId +
// getPlainData at
// serializer-deserializer/src/main/java/com/amazonaws/services/schemaregistry/deserializers/GlueSchemaRegistryDeserializerDataParser.java:67-84
// and 128-161.
//
// Returns ErrIncompatibleData wrapped via fmt.Errorf for:
//   - data shorter than 18 bytes
//   - version byte != 0x03
//   - compression byte not in {0x00, 0x05}
//
// payload is left compressed if compressionByte == 0x05 — the caller decides
// whether to decompress. Returning the raw payload + compression byte keeps
// the wire-format module ignorant of zlib (the CompressionHandler interface
// owns that).
func DecodeWireFormat(data []byte) (schemaVersionID string, compressionByte byte, payload []byte, err error) {
	if len(data) < WireFormatHeaderSize {
		return "", 0, nil, fmt.Errorf("%w: size %d below minimum %d", ErrIncompatibleData, len(data), WireFormatHeaderSize)
	}

	if data[0] != WireFormatVersionByte {
		return "", 0, nil, fmt.Errorf("%w: header version byte %#x != 0x03", ErrIncompatibleData, data[0])
	}

	compressionByte = data[1]
	if compressionByte != CompressionByteNone && compressionByte != CompressionByteZlib {
		return "", 0, nil, fmt.Errorf("%w: compression byte %#x not in {0x00, 0x05}", ErrIncompatibleData, compressionByte)
	}

	// Bytes 2..17 are the UUID — Java writes putLong(MSB) then putLong(LSB),
	// both big-endian, which is byte-identical to a UUID's natural network-order
	// representation. uuid.FromBytes accepts that layout directly.
	var u uuid.UUID
	copy(u[:], data[2:WireFormatHeaderSize])
	schemaVersionID = u.String()

	payload = data[WireFormatHeaderSize:]
	return schemaVersionID, compressionByte, payload, nil
}

File: multilang-schema-registry/typescript/packages/core/src/wire/wire-format.ts
/**
 * 18-byte wire-format header encode/decode.
 *
 * Layout (Java canonical `SerializationDataEncoder.write`, mirrored in Go
 * `core/wire_format.go`):
 *   byte  0:       0x03                                       (version)
 *   byte  1:       0x00 (none) | 0x05 (zlib)                  (compression)
 *   bytes 2..9:    schema-version UUID MSB, big-endian        (8 bytes)
 *   bytes 10..17:  schema-version UUID LSB, big-endian        (8 bytes)
 *   bytes 18..:    payload (already-compressed if byte 1 == 0x05)
 *
 * The MSB-then-LSB big-endian layout is byte-identical to a UUID's natural
 * network-order representation, so parsing the 16 bytes as hex groups
 * 8-4-4-4-12 reproduces the canonical UUID string form.
 *
 * The payload is transparent to this module — compression, protobuf
 * message-index prefixing, and any other transforms happen in the modules
 * that own those concerns and hand the assembled payload to this codec.
 */

import {
  COMPRESSION_BYTE_NONE,
  COMPRESSION_BYTE_ZLIB,
  SCHEMA_VERSION_ID_SIZE,
  WIRE_FORMAT_HEADER_SIZE,
  WIRE_FORMAT_VERSION_BYTE,
} from "./constants.js";
import { GsrIncompatibleDataError } from "./errors.js";

/** Decoded 18-byte header result. */
export interface DecodedWire {
  /** Canonical lowercase UUID string, e.g. `01020304-0506-0708-090a-0b0c0d0e0f10`. */
  schemaVersionId: string;
  /** Raw compression byte from the header (0x00 or 0x05). */
  compressionByte: number;
  /** Payload bytes; left compressed if `compressionByte === 0x05`. */
  payload: Buffer;
}

// Canonical UUID string: 8-4-4-4-12 hex groups, case-insensitive.
const UUID_STRING_REGEX =
  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;

function uuidStringToBytes(schemaVersionId: string): Buffer {
  if (typeof schemaVersionId !== "string" || !UUID_STRING_REGEX.test(schemaVersionId)) {
    throw new GsrIncompatibleDataError(
      `schema version id is not a valid UUID: ${String(schemaVersionId)}`,
    );
  }
  const hex = schemaVersionId.replace(/-/g, "");
  const bytes = Buffer.from(hex, "hex");
  // Buffer.from(hex, "hex") silently drops non-hex chars; the regex above
  // already rejects those, so this is a belt-and-braces guard.
  if (bytes.length !== SCHEMA_VERSION_ID_SIZE) {
    throw new GsrIncompatibleDataError(
      `schema version id did not decode to ${SCHEMA_VERSION_ID_SIZE} bytes: ${schemaVersionId}`,
    );
  }
  return bytes;
}

function uuidBytesToString(data: Buffer, offset: number): string {
  const hex = data.subarray(offset, offset + SCHEMA_VERSION_ID_SIZE).toString("hex");
  return (
    hex.slice(0, 8) +
    "-" +
    hex.slice(8, 12) +
    "-" +
    hex.slice(12, 16) +
    "-" +
    hex.slice(16, 20) +
    "-" +
    hex.slice(20, 32)
  );
}

function assertKnownCompressionByte(compressionByte: number): void {
  if (
    compressionByte !== COMPRESSION_BYTE_NONE &&
    compressionByte !== COMPRESSION_BYTE_ZLIB
  ) {
    throw new GsrIncompatibleDataError(
      `unsupported compression byte 0x${compressionByte.toString(16).padStart(2, "0")}`,
    );
  }
}

/**
 * Assemble the 18-byte wire-format header followed by `payload`.
 *
 * The caller is responsible for producing `payload` in its final form —
 * this module does not compress and does not prepend a protobuf
 * message-index. `schemaVersionId` is caller-supplied so a fixed UUID can
 * be reproduced offline against golden vectors without any live registry
 * interaction.
 *
 * Throws {@link GsrIncompatibleDataError} when the compression byte is not
 * `0x00` or `0x05`, or when the UUID string is malformed.
 */
export function encodeWireHeader(
  schemaVersionId: string,
  compressionByte: number,
  payload: Buffer,
): Buffer {
  assertKnownCompressionByte(compressionByte);
  const uuidBytes = uuidStringToBytes(schemaVersionId);

  const out = Buffer.allocUnsafe(WIRE_FORMAT_HEADER_SIZE + payload.length);
  out[0] = WIRE_FORMAT_VERSION_BYTE;
  out[1] = compressionByte;
  uuidBytes.copy(out, 2);
  payload.copy(out, WIRE_FORMAT_HEADER_SIZE);
  return out;
}

/**
 * Parse the 18-byte wire-format header from `data` and return its fields
 * plus the raw payload slice (payload is left compressed if the header
 * says so; the caller decides whether to decompress).
 *
 * Throws {@link GsrIncompatibleDataError} for data shorter than 18 bytes,
 * a version byte other than `0x03`, or a compression byte not in
 * `{0x00, 0x05}`.
 */
export function decodeWireHeader(data: Buffer): DecodedWire {
  if (data.length < WIRE_FORMAT_HEADER_SIZE) {
    throw new GsrIncompatibleDataError(
      `data length ${data.length} is below minimum header size ${WIRE_FORMAT_HEADER_SIZE}`,
    );
  }

  const versionByte = data[0];
  if (versionByte !== WIRE_FORMAT_VERSION_BYTE) {
    throw new GsrIncompatibleDataError(
      `unsupported wire-format version byte 0x${(versionByte ?? 0).toString(16).padStart(2, "0")}`,
    );
  }

  const compressionByte = data[1] as number;
  assertKnownCompressionByte(compressionByte);

  const schemaVersionId = uuidBytesToString(data, 2);
  const payload = data.subarray(WIRE_FORMAT_HEADER_SIZE);

  return { schemaVersionId, compressionByte, payload };
}
